sycl/doc/KernelProgramCache.md

```diff
@@ -68,42 +68,87 @@ times, which may involve JIT compilation and take quite a lot of time.
 object corresponding to a SPIRV module or native binary defining a set of SYCL
 kernels and/or device functions.
 
 
 ## Data structure of cache
+The cache is split into two levels:
+ - in-memory cache which is used during application runtime for device code
+ which has been already loaded and built for target device.
+ - on-disk cache which is used to store device binaries between application
+ executions.
 
+### In-memory cache
 The cache stores underlying PI objects behind `cl::sycl::program` and
 `cl::sycl::kernel` user-level objects in a per-context data storage. The storage
 consists of two maps: one is for programs and the other is for kernels.
 
-The programs map's key consists of three components: kernel set id<sup>[1](#what-is-ksid)</sup>,
-specialized constants, device this program is built for.
+The programs map's key consists of four components: kernel set id<sup>[1](#what-is-ksid)</sup>,
+specialized constants, device this program is built for, build options id <sup>[2](#what-is-bopts)</sup>.
 
-The krnels map's key consists of three components too: program the kernel
-belongs to, kernel name<sup>[2](#what-is-kname)</sup>, device the program is
-built for.
+The kernels map's key consists of four components too: program the kernel
+belongs to, kernel name<sup>[3](#what-is-kname)</sup>, device the program is
+built for, build options id<sup>[2](#what-is-bopts)</sup>.
 
 <a name="what-is-ksid">1</a>: Kernel set id is an ordinal number of the device
 binary image the kernel is contained in.
-
-<a name="what-is-kname">2</a>: Kernel name is a kernel ID mangled class' name
+<a name="what-is-bopts">2</a>: Hash for the string representation of build
+options set in application or environment variables (e.g.
+SYCL_PROGRAM_COMPILE_OPTIONS, SYCL_PROGRAM_LINK_OPTIONS). The key is hash out
+of the string which is concatenation of `program.get_compile_options()`,
+`program.get_link_options()`, `program.get_build_options()` values.
+<a name="what-is-kname">3</a>: Kernel name is a kernel ID mangled class' name
 which is provided to methods of `cl::sycl::handler` (e.g. `parallel_for` or
 `single_task`).
 
+### On-disk cache
+The cache is hidden behind in-memory cache and stores the same underlying PI
+object behind `cl::sycl::program` user-level objects in a per-context data
+storage.
+The storage consists of the map for storing programs. It uses different keys to
+address difference in SYCL objects ids between applications runs as well as the
+fact that the same kernel name can be used in different SYCL applications.
+
+The programs map's key consists of four components: device image id<sup>[1](#what-is-diid)</sup>,
+specialized constants, device id<sup>[2](#what-is-did)</sup> this program is
+built for, build options id<sup>[3](#what-is-bopts)</sup>.
+<a name="what-is-diid">1</a>: Hash out of  first 10 kB (for performance reasons) of SPIRV image (?and/or source code?) used as input for the build.
```

Comment:
- again, hash equivalence does not guarantee program code equivalence, the program code itself should be part of the key and the hash can be used for fast bail-out. Maybe binary size would be even faster. Or this should be a multi-map.
- would be good not to limit program code to SPIRV form only - can't it be just opaque binary chunk?

Reply:
On-disk cache does not make sense for AOT because no JIT compilation happens in this case. I removed SPIRV wording in favor of device kernel image.

Reply:
> would be good not to limit program code to SPIRV form only - can't it be just opaque binary chunk?

Is it possible to have non-SPIRV binary image for JIT? I believe, it's possible for CUDA, though double checking.

Reply:
My point is: if there are benefits in limiting device binary to SPIRV, they need to be identified and analyzed, otherwise, why limit the design?